### ISSUE TITLE
Log RequestId or Custom Headers

### ISSUE THREAD

**Issue by garylundie** [gogatekeeper/gatekeeper#192]
# Log RequestId or Custom Headers

## Summary 
Provide an option to log the request header (e.g. X-Request-Id) or a custom header (e.g. X-User-Session-Id) in log messages written by gatekeeper

## Why?
When triaging issues across microservices where gatekeeper is used as a sidecar, it is not always clear what logs in gatekeeper relate to the incoming http request as recorded by the upstream server.  As an example, I have recently attempted to trace an error where we see a "405 Method not allowed" in the upstream application server, which I believe is related to an authentication issue, however, I am unable to track the appropriate logs in Gatekeeper, nor am I able to reproduce the issue.

## How
* Update the loggingMiddleware to include the header defined in r.RequestIDHeader
* Expose new config to allow a list of custom headers to be added to the logs
* Update loggingMiddleware and logging for authentication handlers to include r.RequestIDHeader and each header as defined in the newly exposed config

## Acceptance criteria
GIVEN a set of custom logging headers is defined in gatekeeper config
WHEN a request is received by Gatekeeper
THEN all log statements relating to the handling of that request contain the name / value pairs for the configured headers

**Comment by p53** [comment]
Right now logging middleware logs incoming request so what we can log relatively easily are incoming request headers, so if e.g. nginx/apache generates request id it can be logged. But there is option ``enable-request-id`` which can generate uuid for request and add it to header specified in ``request-id-header`` but it will be not in gogatekeeper logs, also custom headers created by gogatekeeper won't be in logs, that would need logging headers etc...also on outgoing request. So i assume you mainly need to log headers from incoming request?

**Comment by p53** [comment]
maybe it would be to log both incoming and outgoing but just in debug level

**Comment by garylundie** [comment]
I think initially the request was to add logging for incoming request headers - allowing the header name to be specified in config as its already in use across other components in the stack.  Reviewing the logging middleware though, I wasn't sure if the incoming request is logged independently, or if its logged along with the other messages - for example, I can see in the logs just now that a request is received, but the headers for that request are not included with subsequent log statements, for example where there is no user session found.  Ideally, I would for the specified header value to be tracked in each JSON logging object written throughout that particular thread - allowing for more detailed tracking of the request.

**Comment by p53** [comment]
probably easiest to do this would be to use enable-request-id which would be generated by gatekeeper for request and log this in other places through entire lifecycle of request but i need to check what that would require

**Comment by garylundie** [comment]
Whilst this does make sense if gatekeeper was the first hop in the journey, the purpose of these headers is track actions originating at a user browser and being passed between multiple microservices.  Right now, our java apps allow us to log the header on each service, as does our proxies, meaning we can query the request ID in our logging stack and see the full request from browser start, even if it passes through multiple components.  However, where there is an error and we track it back to gatekeeper blocking a request, this is difficult to diagnose because we cannot determine which log entries at the timestamp belong to the incoming request.

In short, we can't use the "enable-request-id" as this uses a specific header name (or generates) and means the value would be different to that tracked in the rest of the stack.

**Comment by p53** [comment]
yeah i understand what you mean, that's why i mentioned that we need to log incoming headers also but not all on places, e.g. one log entry would have both incoming headers + gogatekeeper uuid but other gatekeeper entries only gatekeeper uuid, it would be maybe a little bit harder to set up tracking that's true, anyway i will check and think about both aproaches

**Comment by p53** [comment]
@garylundie i enabled logging of headers in request chain (in authentication, authorization, logging middleware), you can enable it by using ``enable-logging=true`` and ``verbose=true``, there is 1.6.1-rc1 image
